### DISCUSSION TITLE
API endpoint to update snake details

### DISCUSSION THREAD

**Discussion by chillfox** [BattlesnakeOfficial/feedback#218] [Feature Requests]
It would be great if we could update a snake's URL with an API call. This would be useful for changing IP address and port during deploys a deploy.

**Comment by robbles** [comment]
Thanks for the suggestion @chillfox!

This is something we're interested in doing, but we also want to make sure we're designing any APIs we expose properly and securely. We probably won't do a one-off endpoint as a beta feature without building out a solid basis for a Battlesnake API, with API tokens, rate limiting, etc.

There's also a more general problem to solve here, of availability/presence management, where we make it easier for Battlesnakes to be offline, or intermittently available without having to constantly ping them. Some ideas that have come up previously:

- An optional endpoint that your server can either ping periodically or maintain a connection to that marks it as "online", and will control whether matchmaking is active for your Battlesnake
- An optional field you could return from your index URL, with a "game URL" that is used instead of the primary one when actually running games
- Options for refreshing your Battlesnake on demand or more often

Any feedback you have on these or other options would be great!

**Reply by coreyja** [reply]
This is kinda out in left field, but I'll share anyways!

I've been doing some Bot dev for Discord and one thing I really like is the ability to do WebSockets for events instead of webhooks. And that way you don't need to be publically accesible on the internet to get events.

I wonder if something similar could be used for Battlesnake! Obv it would be very different and maybe wouldn't interop with HTTP snakes, but I can imagine a Snake calling the engine to connect and open a WebSocket. And from then it just sends game data as websocket messages, and you respond with your moves!

And that gives you 'free' presence, cause if the WebSocket isn't connected the snake isn't around anymore. 
That could also make it easy-ish to add new 'updates' like this as different message types
[Though as I type it out, super long lived web-sockets connections like that would probably be hard on the server but 🤷  Thats an engineering problem to solve 😉 ]

**Reply by robbles** [reply]
It's a really interesting idea - might solve several of our common issues with the HTTP API while creating some new unexpected ones 😆 
There probably is an upper limit on how many Battlesnakes we could support using this at once, but like you said, it's mostly just another engineering problem to solve.

One other concern I have is whether Websockets are as easy to get started with as a plain HTTP web server. I guess if you no longer need to host it on the Internet, it's a tradeoff in complexity? I think I would still want some way to provide the same kind of presence feature to non-Websocket servers if possible.

**Reply by chillfox** [reply]
Adding a game URL field to the index route is an interesting way of solving it. I feel like that is both easier and harder at the same time. So easier because you don't have to figure out how to do API calls with auth, and harder because now you have to run two services, or listen on multiple IPs/Ports.

Just exploring this game URL idea a bit, so you could make it a hash instead of a string and provide a different URL for different sources, say one for battlegrounds and a different one for the league. Maybe even a public URL, so you could run a separate instance for public games that way you can have a snake be public and not impact your league performance. What I am thinking here is on the website, battlegrounds, or league ladder, you see a snake and if it has a public URL specified then there will be a link/button to fight that snake which opens up a new game with that snake added and you can select one of your snakes to fight it. I know people can/do already run a public instance of their snake under a different name, but this would solve the discoverability problem and it kinda would be cool to pick a snake from the ladder to challenge.

Alternatively, having a proper API with auth and all that stuff that is not required for competing but enables nice-to-have functionality would be good as a skills development/learning opportunity. Just add onto Battlesnake being a great fun way to learn modern back-end web development.

Now, onto WebSockets.
I like this idea and think it would be really cool as an option. But the HTTP API should stay available as it's much easier for beginners to get started with.
